cmd/kubectl-gadget/utils/trace.go

```diff
@@ -463,95 +466,157 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
-			nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
+	tracesNumber := len(traceList.Items)
+
+	// We only watch the traces which did not satisfy conditionFunction.
+	tracesToWatch := tracesNumber - len(satisfiedTraces)
+	if tracesToWatch > 0 {
+		// We will need to watch events on them.
+		// For this, we will get a watcher on all the traces which share the same
+		// ID.
+		// Get a watcher on all the traces which have the same ID.
+		// Indeed, all the traces on different nodes but linked to one gadget share
+		// the same ID.
+		// We will also begin to monitor events since the above GET of the traces
+		// list.
+		watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
+		if err != nil {
+			return nil, err
 		}
 
-		if trace.Status.OperationError != "" {
-			watchedTracesNumber++
+		ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
+		_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
+			// This function will be executed until:
+			// 1. The number of watched traces equals the number of traces to watch,
+			// i.e. we dealt with the traces which interest us.
+			// 2. Or it returns an error.
+			// 3. Or time out is fired.
+			// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
+			// with.
+			switch event.Type {
+			case watch.Deleted:
+				// If for some strange reasons (e.g. users deleted a trace during this
+				// operation) a trace is deleted, we need to take care of this by
+				// decrementing the tracesToWatch.
+				// Otherwise we would still wait for the old number and we would
+				// timeout.
+				tracesToWatch--
+
+				trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+				if !ok {
+					return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+				}
 
-			nodeErrors[trace.Spec.Node] = trace.Status.OperationError
+				traceName := trace.ObjectMeta.Name
+
+				// We also remove it from the maps to avoid returning a deleted trace
+				// and timeing out.
+				delete(satisfiedTraces, traceName)
+				delete(erroredTraces, traceName)
+
+				return false, nil
+			case watch.Modified:
+				// We will deal with this type of event below
+			case watch.Error:
+				// Deal particularly with error.
+				return false, fmt.Errorf("Received event is an error one: %v", event)
+			case watch.Added:
+				// createTraces() creates traces synchronously.
+				// So, if a watch.Added event occurs it means there is a problem (e.g.
+				// the user creates a trace by snooping on the traceID of existing
+				// traces).
+				return false, fmt.Errorf("No traces with the given traceID (%s) should be created", traceID)
+			default:
+				// We are not interested in other event types.
+				return false, nil
+			}
 
-			return false, nil
-		}
+			trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+			if !ok {
+				return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+			}
```

Comment:
same here, this only be executed for `watch.Modified` so the assertion is not required here.